src/data/string/ustring.h

```diff
@@ -79,17 +79,21 @@ namespace detail {
 template <class InputIterator1, class InputIterator2>
 uchar chars_to_uchar_impl(InputIterator1& in, InputIterator2 end)
 {
-  if (in == end)
-    throw std::invalid_argument("Empty string");
+  if (in == end) {
+    throw std::invalid_argument("Invalid UTF-8: UTF-8 byte sequences are empty");
+  }
 
   if (((*in) & 0x80) == 0) // U+0000 to U+007F
     return *in++;
 
   const unsigned c = *in & 0xFF;
-  if (c < 0xC2)
-    throw std::invalid_argument("Invalid UTF-8");
-  if (c > 0xFD)
-    throw std::invalid_argument("Invalid UTF-8");
+  // It is an invalid byte also when c is 0xC0 or 0xC1.
+  // But it could only be used for an overlong encoding of ASCII characters,
+  // so it will be checked together after.
+  if (c < 0xC0 || c > 0xFD) {
+    throw std::invalid_argument("Invalid UTF-8: UTF-8 single byte character is out of range. "
```

Comment:
I think, it's not "single byte character" but "first byte of character" is right, since single byte char is already checked by `(((*in) & 0x80) == 0)`